libraries/ms-common/src/stm32f0xx/soft_timer.c

```diff
@@ -0,0 +1,95 @@
+#include "soft_timer.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+
+#include "status.h"
+#include "stm32f0xx_rcc.h"
+#include "stm32f0xx_tim.h"
+
+typedef struct timer {
+  uint32_t duration;
+  uint32_t time_started;
+  soft_timer_callback callback;
+  bool inuse;
+  bool rolled_over;
+  void *context;
+} timer;
+
+static timer s_soft_timer_array[SOFT_TIMER_NUM];
+
+// Used to track rollovers this will happen every ~1.19 hrs as we are using microseconds.
+static uint32_t s_soft_timer_last_value = 0;
+static bool s_soft_timer_initialized = false;
+
+void soft_timer_init(uint8_t clock_speed) {
+  if (!s_soft_timer_initialized) {
+    for (uint8_t i = 0; i < SOFT_TIMER_NUM; i++) {
+      s_soft_timer_array[i].inuse = false;
+    }
+
+    // Configure each clock tick to be 1 microsecond from 0 to UINT32_MAX. Uses the external 8 MHz
+    // clock for tick source. This will need to adjusted or made into a variable if we choose a
+    // different external clock from the discovery board.
+    TIM_TimeBaseInitTypeDef init_struct = {.TIM_Prescaler = clock_speed,
+                                           .TIM_CounterMode = TIM_CounterMode_Up,
+                                           .TIM_Period = UINT32_MAX,
+                                           .TIM_ClockDivision = TIM_CKD_DIV1 };
+
+    RCC_APB1PeriphClockCmd(RCC_APB1Periph_TIM2, ENABLE);
+    TIM_TimeBaseInit(TIM2, &init_struct);
+    TIM_Cmd(TIM2, ENABLE);
+
+    s_soft_timer_initialized = true;
+  }
+}
+
+StatusCode soft_timer_start(uint32_t duration_micros, soft_timer_callback callback,
+                            uint8_t *timer_id, void *context) {
+  uint32_t curr_time = TIM_GetCounter(TIM2);
+  for (uint8_t i = 0; i < SOFT_TIMER_NUM; i++) {
```

Comment:
If you wanted to be really fancy, you could use a bitset instead of tracking timer use within the timer object :P
Not sure if it's worth it in this case, but something to think about.